Sources/Sentry/SentryProfiler.mm

```diff
@@ -370,56 +421,30 @@ + (SentryEnvelopeItem *)createProfilingEnvelopeItemForTransaction:(SentryTransac
 
     // add the gathered metrics
-    NSMutableDictionary<NSString *, id> *metrics = [_gCurrentProfiler->_metricProfiler serialize];
-
-    const auto slicedMetrics =
-        [NSMutableDictionary<NSString *, id> dictionaryWithCapacity:metrics.count];
-    void (^metricSlicer)(NSString *_Nonnull, NSDictionary<NSString *, id> *_Nonnull, BOOL *_Nonnull)
-        = ^(NSString *_Nonnull metricKey, NSDictionary<NSString *, id> *_Nonnull nextMetricsEntry,
-            BOOL *_Nonnull metricStop) {
-              NSArray<NSDictionary<NSString *, NSString *> *> *nextMetrics
-                  = nextMetricsEntry[@"values"];
-
-              const auto nextSlicedMetrics = [self slicedArray:nextMetrics transaction:transaction];
-              if (!nextSlicedMetrics) {
-                  return;
-              }
-
-              slicedMetrics[metricKey] =
-                  @{ @"unit" : nextMetricsEntry[@"unit"], @"values" : nextSlicedMetrics };
-          };
-    [metrics enumerateKeysAndObjectsUsingBlock:metricSlicer];
+    const auto metrics = [_gCurrentProfiler->_metricProfiler serializeForTransaction:transaction];
 
 #    if SENTRY_HAS_UIKIT
-    const auto slowFrames
-        = processFrameRenders(_gCurrentFramesTracker.currentFrames.slowFrameTimestamps,
-            _gCurrentProfiler->_startTimestamp, profileDuration);
+    const auto slowFrames = processFrameRenders(
+        _gCurrentFramesTracker.currentFrames.slowFrameTimestamps, transaction);
     if (slowFrames.count > 0) {
-        slicedMetrics[@"slow_frame_renders"] = @{ @"unit" : @"nanosecond", @"values" : slowFrames };
+        metrics[@"slow_frame_renders"] = @{ @"unit" : @"nanosecond", @"values" : slowFrames };
     }
 
-    const auto frozenFrames
-        = processFrameRenders(_gCurrentFramesTracker.currentFrames.frozenFrameTimestamps,
-            _gCurrentProfiler->_startTimestamp, profileDuration);
+    const auto frozenFrames = processFrameRenders(
+        _gCurrentFramesTracker.currentFrames.frozenFrameTimestamps, transaction);
     if (frozenFrames.count > 0) {
-        slicedMetrics[@"frozen_frame_renders"] =
-            @{ @"unit" : @"nanosecond", @"values" : frozenFrames };
+        metrics[@"frozen_frame_renders"] = @{ @"unit" : @"nanosecond", @"values" : frozenFrames };
     }
 
     const auto frameRates
-        = processFrameRates(_gCurrentFramesTracker.currentFrames.frameRateTimestamps,
-            _gCurrentProfiler->_startTimestamp);
+        = processFrameRates(_gCurrentFramesTracker.currentFrames.frameRateTimestamps, transaction);
     if (frameRates.count > 0) {
-        const auto slicedFrameRates = [self slicedArray:frameRates transaction:transaction];
-        if (slicedFrameRates.count > 0) {
-            slicedMetrics[@"screen_frame_rates"] =
-                @{ @"unit" : @"hz", @"values" : slicedFrameRates };
-        }
+        metrics[@"screen_frame_rates"] = @{ @"unit" : @"hz", @"values" : frameRates };
```

Comment:
The slicing of frame rate entries is now done directly in processFrameRates with a couple timestamp checks.